src/hammer-vlsi/hammer_vlsi/constraints.py

```diff
@@ -242,42 +244,138 @@ class PlacementConstraint(NamedTuple('PlacementConstraint', [
         if "orientation" in constraint:
+            assert constraint_type != PlacementConstraintType.TopLevel, "Non-TopLevel constraint must not contain orientation: {}".format(constraint)
             orientation = str(constraint["orientation"])
+
+        ### Top layer ###
+        # This field is optional in HardMacro constraints
+        # This field is disallowed otherwise
+        top_layer = None  # type: Optional[str]
         if "top_layer" in constraint:
+            assert constraint_type == PlacementConstraintType.HardMacro, "Non-HardMacro constraint must not contain top_layer: {}".format(constraint)
             top_layer = str(constraint["top_layer"])
+
+        ### Layers ###
+        # This field is optional in Obstruction constraints
+        # This field is disallowed otherwise
+        layers = None  # type: Optional[List[str]]
         if "layers" in constraint:
+            assert constraint_type == PlacementConstraintType.Obstruction, "Non-Obstruction constraint must not contain layers: {}".format(constraint)
             layers = []
             for layer in constraint["layers"]:
                 layers.append(str(layer))
+
+        ### Obstruction types ###
+        # This field is mandatory in Obstruction constraints
+        # This field is disallowed otherwise
+        obs_types = None  # type: Optional[List[ObstructionType]]
         if "obs_types" in constraint:
+            assert constraint_type == PlacementConstraintType.Obstruction, "Non-Obstruction constraint must not contain obs_types: {}".format(constraint)
             obs_types = []
             types = constraint["obs_types"]
             for obs_type in types:
                 obs_types.append(ObstructionType.from_str(str(obs_type)))
+        else:
+            assert constraint_type != PlacementConstraintType.Obstruction, "Obstruction constraint must contain obs_types: {}".format(constraint)
+
+        ### Master ###
+        # This field is mandatory for Hierarchical constraints
+        # This field is optional for HardMacro constraints
+        # This field is disallowed otherwise
+        master = None  # type: Optional[str]
+        if "master" in constraint:
+            assert constraint_type in [PlacementConstraintType.Hierarchical, PlacementConstraintType.HardMacro], "Constraints other than Hierarchical and HardMacro must not contain master: {}".format(constraint)
+            master = constraint["master"]
+        else:
+            assert constraint_type != PlacementConstraintType.Hierarchical, "Hierarchical constraint must contain master: {}".format(constraint)
+
+        ### Width & height ###
+        # These fields are mandatory for Dummy, Placement, TopLevel, and Obstruction constraints
+        # These fields are optional for HardMacro and Hierarchical constraints
+        #   If omitted, they are copied from the master definition (HardMacro) or hierarchical top-level width and height (Hierarchical).
```

Comment:
Not sure I followed what you meant here by `master` (see above)

Reply:
`master` is the module or macro name (that's the reason for the addition of the `master` field in the tuple). Once that is documented, I think it'll be clearer.